src/Features/Core/Portable/PullMemberUp/CodeActionAndSolutionGenerator.cs

```diff
@@ -0,0 +1,409 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeGeneration;
+using Microsoft.CodeAnalysis.CodeRefactorings.PullMemberUp;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Formatting;
+using Roslyn.Utilities;
+using static Microsoft.CodeAnalysis.CodeActions.CodeAction;
+
+namespace Microsoft.CodeAnalysis.PullMemberUp
+{
+    internal class CodeActionAndSolutionGenerator
+    {
+        internal async Task<Solution> GetSolutionAsync(
+            AnalysisResult result,
+            Document contextDocument,
+            CancellationToken cancellationToken)
+        {
+            var codeGenerationService = contextDocument.Project.LanguageServices.GetRequiredService<ICodeGenerationService>();
+            var solution = contextDocument.Project.Solution;
+            var targetNodeSyntax = await codeGenerationService.FindMostRelevantNameSpaceOrTypeDeclarationAsync(solution, result.Target).ConfigureAwait(false);
+
+            if (targetNodeSyntax != null)
```

Comment:
when is the containing function being called wrt the user interacting with this feature?  Is this called when the user chooses the item from the lightbulb?  If so, this isn't ok that we could have a null target, since that means the user will select things and we'll do nothing.  We need to validate before showing them anything that we'll actually be able to make the change.